[Metricbeat] 7.4.1 raises a panic error with Kubernetes 1.16 and Deployment

- Version: 7.4.1 (it works with 7.3.2 in the same environment and with the same manifest)
- Operating System: Ubuntu and Kubernetes 1.16.2
- Issue description:

Trying to make the default manifest from the doc [running metricbeat in Kubernetes](https://www.elastic.co/guide/en/beats/metricbeat/7.4/running-on-kubernetes.html) running in Kubernetes 1.16.2 I have found a weird crash.

The official manifest provides a DaemonSet + Deployment. The DaemonSet with 7.4.1 image works fine, but the Deployment crashes with a `panic` error claiming `invalid memory address or nil pointer dereference`.

Just changing the image to `7.3.2` makes the manifest to work (not exactly the original manifest,  I had to do small changes  to make it compatible with Kubernetes 1.16). Feel  free  to consider these changes for the official manifest for 1.16 clusters btw.

- Configuration used:
[metricbeat-kubernetes-kube-system-sharing.yaml.zip](https://github.com/elastic/beats/files/3786235/metricbeat-kubernetes-kube-system-sharing.yaml.zip)

-  Changes highlighted:
   - `selector --> matchLabels` is  mandatory in kubernetes 1.16.2, so I  had to add them at both DaemonSet and Deployment levels (this is already included in the official manifest in `master` branch).
   - Don't really know why, but I had to add `-replicasets` in the `ClusterRole` definition. That's not included in the `master` manifest and I don't know if I had to add it due to the specifics of  my Kubernetes installation, or if for some reason is a real need of Metricbeat now.

```
- apiGroups: ["apps"]
  resources:
  - statefulsets
  - deployments
  - replicasets
  verbs: ["get", "list", "watch"]
```

- I also had to add `labels.dedot` and `annotations.dedot` options in the Deployment configuration, otherwise I had a lot of mapping errors when running the 7.3.1 image.

In summary:
- Running metricbeat with the provided config file works with image 7.3.2.
- If image 7.4.1 is used the deployment always crashes (it doesn't matter if the fixes (dedot, clusterRole, etc) are applied or not.

- Logs of the crash:
```
2019-10-29T23:00:22.844Z	INFO	instance/beat.go:607	Home path: [/usr/share/metricbeat] Config path: [/usr/share/metricbeat] Data path: [/usr/share/metricbeat/data] Logs path: [/usr/share/metricbeat/logs]
2019-10-29T23:00:22.850Z	INFO	instance/beat.go:615	Beat ID: b8a0232c-ad4f-422b-80b2-0fbea618060f
2019-10-29T23:00:22.852Z	INFO	[seccomp]	seccomp/seccomp.go:124	Syscall filter successfully installed
2019-10-29T23:00:22.852Z	INFO	[beat]	instance/beat.go:903	Beat info	{"system_info": {"beat": {"path": {"config": "/usr/share/metricbeat", "data": "/usr/share/metricbeat/data", "home": "/usr/share/metricbeat", "logs": "/usr/share/metricbeat/logs"}, "type": "metricbeat", "uuid": "b8a0232c-ad4f-422b-80b2-0fbea618060f"}}}
2019-10-29T23:00:22.852Z	INFO	[beat]	instance/beat.go:912	Build info	{"system_info": {"build": {"commit": "12ee6cd05c1bfdc69721ddab1f473417b1514576", "libbeat": "7.4.1", "time": "2019-10-22T16:26:05.000Z", "version": "7.4.1"}}}
2019-10-29T23:00:22.852Z	INFO	[beat]	instance/beat.go:915	Go runtime info	{"system_info": {"go": {"os":"linux","arch":"amd64","max_procs":4,"version":"go1.12.9"}}}
2019-10-29T23:00:22.857Z	INFO	[beat]	instance/beat.go:919	Host info	{"system_info": {"host": {"architecture":"x86_64","boot_time":"2019-10-28T21:46:08Z","containerized":false,"name":"worker1","ip":["127.0.0.1/8","::1/128","78.47.135.46/32","78.47.101.55/32","2a01:4f8:c010:4903::1/64","fe80::9400:ff:fe34:692b/64","10.0.2.2/32","fe80::8400:ff:fe34:692d/64","172.17.0.1/16","fe80::4098:d2ff:fe23:c81a/64","192.168.3.17/32","fe80::c05a:ceff:fec6:ecc5/64","fe80::b039:8ff:fe41:d5b7/64","fe80::f8b8:83ff:feaf:7455/64","fe80::60f0:31ff:fe2e:7a80/64","fe80::e8d0:1eff:fe52:d898/64","fe80::205e:f8ff:fe62:19d7/64","fe80::f80d:19ff:fef4:6c71/64","fe80::206c:d7ff:fec9:7256/64","fe80::f0d3:e7ff:fecf:9b94/64","fe80::7c5b:22ff:fe18:8625/64","fe80::242b:6eff:fed9:c385/64","fe80::b005:a1ff:fe5f:88ff/64"],"kernel_version":"4.15.0-64-generic","mac":["96:00:00:34:69:2b","86:00:00:34:69:2d","02:42:e6:82:f7:39","42:98:d2:23:c8:1a","c2:5a:ce:c6:ec:c5","b2:39:08:41:d5:b7","fa:b8:83:af:74:55","62:f0:31:2e:7a:80","ea:d0:1e:52:d8:98","22:5e:f8:62:19:d7","fa:0d:19:f4:6c:71","22:6c:d7:c9:72:56","f2:d3:e7:cf:9b:94","7e:5b:22:18:86:25","26:2b:6e:d9:c3:85","b2:05:a1:5f:88:ff"],"os":{"family":"redhat","platform":"centos","name":"CentOS Linux","version":"7 (Core)","major":7,"minor":6,"patch":1810,"codename":"Core"},"timezone":"UTC","timezone_offset_sec":0}}}
2019-10-29T23:00:22.858Z	INFO	[beat]	instance/beat.go:948	Process info	{"system_info": {"process": {"capabilities": {"inheritable":["chown","dac_override","fowner","fsetid","kill","setgid","setuid","setpcap","net_bind_service","net_raw","sys_chroot","mknod","audit_write","setfcap"],"permitted":["chown","dac_override","fowner","fsetid","kill","setgid","setuid","setpcap","net_bind_service","net_raw","sys_chroot","mknod","audit_write","setfcap"],"effective":["chown","dac_override","fowner","fsetid","kill","setgid","setuid","setpcap","net_bind_service","net_raw","sys_chroot","mknod","audit_write","setfcap"],"bounding":["chown","dac_override","fowner","fsetid","kill","setgid","setuid","setpcap","net_bind_service","net_raw","sys_chroot","mknod","audit_write","setfcap"],"ambient":null}, "cwd": "/usr/share/metricbeat", "exe": "/usr/share/metricbeat/metricbeat", "name": "metricbeat", "pid": 1, "ppid": 0, "seccomp": {"mode":"filter","no_new_privs":true}, "start_time": "2019-10-29T23:00:22.110Z"}}}
2019-10-29T23:00:22.858Z	INFO	instance/beat.go:292	Setup Beat: metricbeat; Version: 7.4.1
2019-10-29T23:00:22.858Z	INFO	[index-management]	idxmgmt/std.go:178	Set output.elasticsearch.index to 'metricbeat-7.4.1' as ILM is enabled.
2019-10-29T23:00:22.859Z	INFO	elasticsearch/client.go:170	Elasticsearch url: https://elastic-k8s-es-http.monitoring:9200
2019-10-29T23:00:22.859Z	INFO	[publisher]	pipeline/module.go:97	Beat name: worker1
2019-10-29T23:00:22.861Z	INFO	add_cloud_metadata/add_cloud_metadata.go:87	add_cloud_metadata: hosting provider type not detected.
2019-10-29T23:00:22.862Z	INFO	instance/beat.go:422	metricbeat start running.
2019-10-29T23:00:22.862Z	INFO	[monitoring]	log/log.go:118	Starting metrics logging every 30s
2019-10-29T23:00:22.870Z	INFO	cfgfile/reload.go:171	Config reloader started
2019-10-29T23:00:22.885Z	INFO	cfgfile/reload.go:226	Loading of config files completed.
E1029 23:00:22.986433       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 296 [running]:
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3fbb0e0, 0x6eebab0)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:69 +0x7b
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51 +0x82
panic(0x3fbb0e0, 0x6eebab0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/elastic/beats/libbeat/common/kubernetes.(*MetaGeneratorConfig).ResourceMetadata(0xc000997d60, 0x4a54f80, 0xc0001f8540, 0x451844f)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/metadata.go:113 +0x485
github.com/elastic/beats/metricbeat/module/kubernetes/util.NewResourceMetadataEnricher.func1(0xc0009dc270, 0x4a54f80, 0xc0001f8540)
	/go/src/github.com/elastic/beats/metricbeat/module/kubernetes/util/kubernetes.go:149 +0x1e9
github.com/elastic/beats/metricbeat/module/kubernetes/util.buildMetadataEnricher.func1(0x44e89a0, 0xc0001f8540)
	/go/src/github.com/elastic/beats/metricbeat/module/kubernetes/util/kubernetes.go:268 +0xb1
github.com/elastic/beats/libbeat/common/kubernetes.ResourceEventHandlerFuncs.OnAdd(...)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/eventhandler.go:54
github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).process.func1(0xc00026f3b0, 0x3d87fe0, 0xc000ec2c20, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:284 +0x1e0
github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).process(0xc00026f3b0, 0x4a92400, 0xc000989bc0, 0x1)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:292 +0x4d
github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).Start.func1()
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:247 +0x48
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0009a8740)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x54
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0009a8740, 0x3b9aca00, 0x0, 0xc0007d7501, 0xc00005a480)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0009a8740, 0x3b9aca00, 0xc00005a480)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).Start
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:246 +0x1ef
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2c21015]

goroutine 296 [running]:
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x3fbb0e0, 0x6eebab0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/elastic/beats/libbeat/common/kubernetes.(*MetaGeneratorConfig).ResourceMetadata(0xc000997d60, 0x4a54f80, 0xc0001f8540, 0x451844f)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/metadata.go:113 +0x485
github.com/elastic/beats/metricbeat/module/kubernetes/util.NewResourceMetadataEnricher.func1(0xc0009dc270, 0x4a54f80, 0xc0001f8540)
	/go/src/github.com/elastic/beats/metricbeat/module/kubernetes/util/kubernetes.go:149 +0x1e9
github.com/elastic/beats/metricbeat/module/kubernetes/util.buildMetadataEnricher.func1(0x44e89a0, 0xc0001f8540)
	/go/src/github.com/elastic/beats/metricbeat/module/kubernetes/util/kubernetes.go:268 +0xb1
github.com/elastic/beats/libbeat/common/kubernetes.ResourceEventHandlerFuncs.OnAdd(...)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/eventhandler.go:54
github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).process.func1(0xc00026f3b0, 0x3d87fe0, 0xc000ec2c20, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:284 +0x1e0
github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).process(0xc00026f3b0, 0x4a92400, 0xc000989bc0, 0x1)
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:292 +0x4d
github.com/elastic/beats/libbeat/common/kubernetes.(*watcher).Start.func1()
	/go/src/github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:247 +0x48
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0009a8740)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x54
github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0009a8740, 0x3b9aca00, 0x0, 0xc0007d7501, 0xc00005a480)
	/go/src/github.com/elastic/beats/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
```

Any idea of what's causing that and if there's an easy fix?

--- a/CHANGELOG.next.asciidoc
+++ b/CHANGELOG.next.asciidoc
@@ -120,6 +120,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Fix panics that could result from invalid TLS certificates. This can affect Beats that connect over
   TLS or Beats that accept connections over TLS and validate client certificates. {pull}14146[14146]
 - Support usage of custom builders without hints and mappers {pull}13839[13839]
+- Fix kubernetes `metaGenerator.ResourceMetadata` when parent reference controller is nil {issue}14320[14320] {pull}14329[14329]
 
 *Auditbeat*
 

--- a/libbeat/common/kubernetes/metadata.go
+++ b/libbeat/common/kubernetes/metadata.go
@@ -110,7 +110,7 @@ func (g *metaGenerator) ResourceMetadata(obj Resource) common.MapStr {
 	// Add controller metadata if present
 	if g.IncludeCreatorMetadata {
 		for _, ref := range accessor.GetOwnerReferences() {
-			if *ref.Controller {
+			if ref.Controller != nil && *ref.Controller {
 				switch ref.Kind {
 				// TODO grow this list as we keep adding more `state_*` metricsets
 				case "Deployment",

--- a/libbeat/common/kubernetes/metadata_test.go
+++ b/libbeat/common/kubernetes/metadata_test.go
@@ -21,7 +21,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 
@@ -37,10 +37,12 @@ func TestPodMetadata(t *testing.T) {
 	True := true
 	False := false
 	tests := []struct {
+		name string
 		pod  *Pod
 		meta common.MapStr
 	}{
 		{
+			name: "standalone Pod",
 			pod: &Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Labels:    map[string]string{"a.key": "foo", "a": "bar"},
@@ -62,6 +64,7 @@ func TestPodMetadata(t *testing.T) {
 			},
 		},
 		{
+			name: "Deployment + Replicaset owned Pod",
 			pod: &Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Labels: map[string]string{"a.key": "foo", "a": "bar"},
@@ -94,6 +97,7 @@ func TestPodMetadata(t *testing.T) {
 			},
 		},
 		{
+			name: "StatefulSet + Deployment owned Pod",
 			pod: &Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Labels: map[string]string{"a.key": "foo", "a": "bar"},
@@ -131,6 +135,29 @@ func TestPodMetadata(t *testing.T) {
 				"statefulset": common.MapStr{"name": "StatefulSet"},
 			},
 		},
+		{
+			name: "empty owner reference Pod",
+			pod: &Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:          map[string]string{"a.key": "foo", "a": "bar"},
+					UID:             types.UID(UID),
+					OwnerReferences: []metav1.OwnerReference{{}},
+					Namespace:       test,
+				},
+				Spec: v1.PodSpec{
+					NodeName: test,
+				},
+			},
+			meta: common.MapStr{
+				"pod": common.MapStr{
+					"name": "",
+					"uid":  "005f3b90-4b9d-12f8-acf0-31020a840133",
+				},
+				"node":      common.MapStr{"name": "test"},
+				"namespace": "test",
+				"labels":    common.MapStr{"a": common.MapStr{"value": "bar", "key": "foo"}},
+			},
+		},
 	}
 
 	for _, test := range tests {
@@ -142,9 +169,9 @@ func TestPodMetadata(t *testing.T) {
 
 		metaGen, err := NewMetaGenerator(config)
 		if err != nil {
-			t.Fatal(err)
+			t.Fatalf("case %q failed: %s", test.name, err.Error())
 		}
-		assert.Equal(t, metaGen.PodMetadata(test.pod), test.meta)
+		assert.Equal(t, metaGen.PodMetadata(test.pod), test.meta, "test failed for case %q", test.name)
 	}
 }
 

